lib/mongo/server/connection.rb

```diff
@@ -173,11 +173,21 @@ def connect!
 
       # Separate method to permit easier mocking in the test suite.
       def do_connect
-        socket = address.socket(socket_timeout, ssl_options,
-          connect_timeout: address.connect_timeout)
-        handshake!(socket)
-        pending_connection = PendingConnection.new(socket, @server, monitoring, options)
-        authenticate!(pending_connection)
+        socket = nil
+
+        begin
+          socket = address.socket(socket_timeout, ssl_options,
+            connect_timeout: address.connect_timeout)
```

Comment:
If you move `begin` below this line, the rescue block won't need to check whether `socket` is not nil.

Reply:
I thought about this, but it looks like the call on this line could also throw an exception, which is why I included it in the block.

Reply:
Yes but if it does, the cleanup code in this method won't do anything since `socket` won't be assigned.